### PR TITLE
feat(extras): added fsharp support

### PR DESCRIPTION
**Pull request by muneebusmani**
Added fsharp support  #3470

### CODE REVIEW

**Review by cjalvis — APPROVED** [review]

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (folke) [inline lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: Would be best to have this in nvim-treesitter first

**Comment by folke** [comment]
Would also be great if you could read https://github.com/LazyVim/LazyVim/blob/main/CONTRIBUTING.md
(if you haven't done so).

The doc is new

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:25]: Yeah it provides the functionality which omnisharp-extended-lsp provides, like full project scanning and completions, which wouldn't work properly without this plugin, because the fsharp_language_server stated in server_configurations in nvim lspconfig is no longer maintained, and is stopped at .net 6 i guess, and current lts is .net 8

**Comment by muneebusmani** [comment]
https://github.com/ionide/Ionide-vim?tab=readme-ov-file#requirements
Check the repo out , this plugin supports both vim and neovim and it is written in lua and vimscript both, to support both editors, and there is no lua only solution for fsharp, it is easily overridable , according to my knowledge

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: It is a custom parser defined above

Comment (amaanq) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: adding the parser is useless without any queries, this should be PRed first to nvim-treesitter if the parser is of good enough quality and can parse most of the F# spec, then can be added here

Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: > adding the parser is useless without any queries, this should be PRed first to nvim-treesitter if the parser is of good enough quality and can parse most of the F# spec, then can be added here

I didn't add any query, this plugin has queries in it's repo

Comment (amaanq) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: queries in its repo =/= queries that neovim supports and/or are in nvim-treesitter

Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: Is there a way to add queries with extras?

Comment (amaanq) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: Not directly, and you shouldn't. Again, it should be added to nvim-treesitter first if the parser is of good enough quality.

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/fsharp.lua:62]: Should this be enabled with the `Ionide-vim`? From its [README](https://github.com/ionide/Ionide-vim?tab=readme-ov-file#for-neovim-05)
> If you are using [neovim/nvim-lspconfig](https://github.com/neovim/nvim-lspconfig), do not enable fsautocomplete. Ionide-vim automatically integrates itself into nvim-lspconfig and will register itself as a server.

Would it be better to hold up on this PR until more people who do f-sharp weigh in their opinions?

Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:62]: My mistake, I didn't read that part in the readme, if you are right, this pr needs to be holded

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:62]: Do you actually program in f-sharp? Or did you only want to provide support in LazyVim for f-sharp because there was a request? This is actually a naive question as I don't know anything about you. But I believe that language Extras should actually be provided by people with experience in the specific language domain.

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: Dude this thing works and that's what ppl care about who are new to neovim and LazyVim, idk if that parser is good or not because I don't do fsharp but have done csharp so I have a general idea of dotnet ecosystem,  there isn't any other person working on fsharp parser so there isn't a choice

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (muneebusmani) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:62]: I did csharp before and know a lil bit of fsharp , and have great idea of dotnet ecosystem so that's why I searched about these things and implemented them

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:62]: My personal opinion is to hold up on this PR since there are quite a few lacking areas that we aren't sure about. I'd prefer to wait for more people who do f-sharp to weigh in their opinions. But in the end it's up to the maintainer to decide.

Comment (folke) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:62]: Good point.
Just updated the CONTRIBUTING doc.

@muneebusmani new languages really should only be added by people that use these languages already and have a lot of experience in configuring the language for Neovim.

You have to understand that when I merge a PR for a new plugin/language, then that also needs to be maintained. To lessen the burden on maintenance, the new lang should at least be configured by someone that knows the config well.

Thank you for the effort, but I'm going to close this for now.

File: lua/lazyvim/plugins/extras/lang/fsharp.lua
Comment (amaanq) [reply lua/lazyvim/plugins/extras/lang/fsharp.lua:22]: > Dude this thing works

It does *not* work, and if you don't do f-sharp then maybe don't add the parser without bothering to understand how tree-sitter/nvim-treesitter work? Adding it here does *absolutely* nothing without proper queries

**Pull request closed without merging**
